raft/server.cc

```diff
@@ -450,6 +468,38 @@ future<> server_impl::wait_for_state_change(seastar::abort_source* as) {
     }
 }
 
+future<server::trigger_snapshot_result> server_impl::trigger_snapshot(seastar::abort_source& as) {
+    if (!_applied_idx) {
+        throw std::runtime_error{
+            "Cannot trigger a snapshot with no commands applied"};
+    }
+
+    if (_fsm->log_last_idx() == _fsm->log_last_snapshot_idx()) {
+        throw std::runtime_error{fmt::format(
+            "Last Raft log index is the same as last snapshot index ({}),"
+            " cannot trigger a snapshot.", _fsm->log_last_idx())};
+    }
+
+    if (_awaited_snapshot) {
+        // Might delay a concurrent `trigger_snapshot` request but that's OK.
+        // Note that min_idx can only grow since _applied_idx can only grow.
```

Comment:
This is exactly why I do not like this `_awaited_snapshot` thing. You can hold `min_idx` here in the fiber frame and check it exactly. You do not need one global `min_idx`. You only need `_snapshot_taken` CV to wait upon.

Reply:
> You can hold min_idx here in the fiber frame and check it exactly.

You need to check it against something --- persisted snapshot descriptor index. We don't have it. So you have to introduce a new thing. I decided to go with the spirit of existing data structures like
```cpp
    std::multimap<index_t, awaited_index> _awaited_indexes;
```
except instead of a map of indexes, I have only one index. Having only one index instead of a map has this minor consequence that one request might delay another. But it's not important consequence. We could also disallow concurrent `trigger_snapshot` but it wouldn't change the number of different data structures.

Reply:
> > You can hold min_idx here in the fiber frame and check it exactly.
> 
> You need to check it against something --- persisted snapshot descriptor index. We don't have it. 

But we can have it as we do for `_applied_idx`. 

> So you have to introduce a new thing. I decided to go with the spirit of existing data structures like
> 
> ```c++
>     std::multimap<index_t, awaited_index> _awaited_indexes;
> ```
> 
> except instead of a map of indexes, I have only one index. Having only one index instead of a map has this minor consequence that one request might delay another. But it's not important consequence. We could also disallow concurrent `trigger_snapshot` but it wouldn't change the number of different data structures.


So we can either store shared_prpomise + and awaited index like you do or we can have a CV and snapshot index like I propose. This is basically the same, but later handles multiple calls naturally. Why not do it?

Reply:
OK, I'll try this. We will lose information whether the snapshot we waited on was local or not and the guarantee that local snapshots give but I currently see no use case for this anyway so it doesn't matter.

Reply:
After I wrote the code I realized there is a bug with this approach....
```cpp
        co_await _snapshot_desc_idx_changed.when([this, &as] {
            return as.abort_requested() || _applied_idx <= _snapshot_desc_idx;
        });
```

Even though a snapshot was created by the request, by the time this `when` wakes up, `_applied_idx` might have already gone past `_snapshot_desc_idx`.

Even better, it might have gone past already at the moment we signal `_snapshot_desc_idx_changed`!

So what do we do, trigger a new snapshot every time we wake up and see the condition is false? Ignoring the fact that it would just trigger multiple snapshots unnecessarily --- even that is not live, as the apply index might constantly move forward and outpace these requests.

Reply:
Ah, I'm stupid. I need to store a local copy of `_applied_idx`.